import { ValidCategory, ValidExpType, ValidSkills } from "./constants";

interface PagesInfoInterface {
  title: string;
  imgArr: string[];
  linkURL?: string;
  description?: string;
}

interface DescriptionDetailsInterface {
  paragraphs: string[];
  bullets: string[];
}

export interface ProjectInterface {
  id: string;
  type: ValidExpType;
  companyName: string;
  category: ValidCategory[];
  shortDescription: string;
  websiteLink?: string;
  githubLink?: string;
  techStack: ValidSkills[];
  startDate: Date;
  endDate: Date;
  companyLogoImg: any;
  descriptionDetails: DescriptionDetailsInterface;
  pagesInfoTitle?: string;
  pagesInfoArr: PagesInfoInterface[];
  link?: string;
}

export const Experiences: ProjectInterface[] = [
  {
    id: "dna-quality-control-toolkit",
    companyName: "DNA QC Toolkit",
    type: "Personal Project",
    category: ["Bioinformatics", "Computational Bioengineering", "UI/UX"],
    shortDescription:
      "Developed and optimized a high-performing quality control toolkit for DNA sequence analysis, emphasizing efficiency and accuracy.",
    websiteLink: "https://github.com/aahmeddev/QC-Toolkit/",
    techStack: ["Java", "Python", "Git", "HTML 5", "JavaFX", "Biopython"],
    startDate: new Date("2025-07-01"),
    endDate: new Date("2025-07-27"),
    companyLogoImg: "/portfolio/projects/dna-quality-control-toolkit/logo.png",
    pagesInfoTitle: "Features",
    pagesInfoArr: [
      {
        title: "Welcome Page",
        description:
          "Select .fastq/fastq.gz file for quality analysis, dragging from file explorer or browsing locally.",
        imgArr: [
          "/portfolio/projects/dna-quality-control-toolkit/welcome_page.png",
        ],
      },
      {
        title: "Base Statistics",
        description:
          "Get a quality overview for the selected tab's data, and the core information, including areas of concern to do with the data.",
        imgArr: [
          "/portfolio/projects/dna-quality-control-toolkit/base_statistics.png",
        ],
      },
      {
        title: "Data Visualisation",
        description:
          "Select between 10 different plots, and learn essential information on the base sequences and dataset information. See these examples from a good quality dataset:",
        imgArr: [
          "/portfolio/projects/dna-quality-control-toolkit/data_examples.png",
        ],
      },
      {
        title: "Data Trimming",
        description:
          "Remove adapter and poor quality sequences, allowing for an overrall better quality dataset",
        imgArr: [
          "/portfolio/projects/dna-quality-control-toolkit/trimming.png",
        ],
      },
      {
        title: "Export to HTML",
        description:
          "Export the report as an interactive HTML, allowing it to be shared and saved.",
        imgArr: [],
        linkURL:
          "/portfolio/projects/dna-quality-control-toolkit/SRR490124_2_quality_report.html",
      },
    ],
    descriptionDetails: {
      paragraphs: [
        "In any DNA sequencing project, the raw data generated by sequencing machines is the foundation for all subsequent discoveries. However, this raw data is rarely perfect. It can contain a range of technical errors and artifacts, such as low-quality reads, leftover adapter sequences from the lab process, and other biases.",
        "This is where a Quality Control (QC) application comes in. It serves as the first and most critical checkpoint in the bioinformatics pipeline. Before any meaningful biological analysis can begin—such as aligning reads to a genome or identifying mutations—researchers must first use a tool like this one to assess the health of their raw data",
        "By generating a comprehensive summary of data quality, these applications allow scientists to diagnose potential issues. Based on the QC report, a researcher can then make an informed decision: either the data is of high quality and ready for analysis, or it needs to be cleaned (e.g., by trimming adapters and low-quality ends) to ensure that the final scientific conclusions are accurate and reliable. To get a deeper understanding of the quality control process, I made an amateur DNA QC Toolkit, inspired by applications like FastQC.",
        "A key part of this project was developing a deep understanding of the FASTQ file format, including how to parse its four-line structure to extract both the DNA sequences and their corresponding Phred quality scores. I researched and implemented the logic for several essential quality control (QC) metrics, such as Per-Base Sequence Quality, GC Content, Sequence Duplication Levels, and Adapter Contamination, ensuring the application provides a scientifically valid analysis.",
        "The underlying application logic, including file I/O and data manipulation, was implemented in Java. For the user interface, JavaFX and FXML were used to create a user-friendly and responsive graphical application with the multi-tabbed design. To prevent UI freezing while performing analysis of large files, I implemented a multi-threaded system based on Java's ExecutorService where all intensive computations are performed on background threads.",
        "All plots are visualized using a Python backend. The central Java application governs and communicates with external Python processes. This option leverages the strengths and versatility of the Matplotlib library to generate the entire range of scientific plots and tables required for final analysis reports.",
      ],
      bullets: [],
    },
  },
  {
    id: "gene-expression-explorer",
    companyName: "Gene Expression Explorer",
    type: "Personal Project",
    category: ["Computational Bioengineering", "Web Dev", "UI/UX"],
    shortDescription:
      "An interactive web app for gene expression data analysis and visualization. Upload your dataset, define sample groups easily, view heatmaps, volcano plots, gene boxplots, and download results—all powered by Streamlit and Python.",
    websiteLink: "https://gene-expression-explorer.streamlit.app/",
    githubLink: "https://github.com/aahmeddev/gene-expression-explorer",
    techStack: ["Python", "Streamlit"],
    startDate: new Date("2025-08-22"),
    endDate: new Date("2022-09-01"),
    companyLogoImg: "/portfolio/projects/gene-expression-explorer/logo.png",
    pagesInfoTitle: "Visit The Streamlit App",
    pagesInfoArr: [
      {
        title: "",
        linkURL: "https://gene-expression-explorer.streamlit.app",
        imgArr: [],
      }
    ],
    descriptionDetails: {
      paragraphs: [
        "In any genomics research project, once sequencing data has passed quality control and been processed, the next critical step is understanding which genes are actually changing between experimental conditions. Raw gene expression data—whether from RNA-seq, microarrays, or other transcriptomic experiments—contains thousands of measurements, but without proper statistical analysis, it's impossible to distinguish meaningful biological signals from technical noise.\n",
        "This is where a differential gene expression analysis tool becomes essential. It serves as the analytical bridge between raw expression measurements and biological insights. Before researchers can make claims about which genes are upregulated in disease versus healthy tissue, or which pathways respond to a particular treatment, they must first use a tool like this one to rigorously identify statistically significant expression changes and filter out genes that vary due to experimental variation.\n",
        "A key part of this process is understanding both the size and the reliability of changes in gene expression. The fold change tells us how many times higher or lower a gene’s expression is in one group compared to another, but to make these changes easier to interpret, we use the log2 fold change. This log scale means that a value of 1 represents a twofold increase, while -1 means a twofold decrease, making it straightforward to compare up- and down-regulation. However, not all observed changes are meaningful as random variation can make a gene appear different when it’s not. That’s why statistical tests, such as the t-test, are used to calculate a p-value, which estimates the likelihood that the observed difference happened by chance. By considering both the magnitude of change (log2 fold change) and its statistical significance (p-value), tools like these can help identify which genes are truly differentially expressed.\n",
        "The data analysis was implemented entirely in Python using pandas for data manipulation, NumPy for numerical computations, and SciPy for statistical testing. For the user interface, I chose Streamlit to create an intuitive, web-based application to easily upload data, and explore results interactively. Users can also configure analysis parameters from the sidebar, featuring real-time feedback on data loading and processing status.\n",
        "All visualizations use both static and interactive plotting libraries. The application generates clustered heatmaps using Seaborn and Matplotlib, while interactive volcano plots and gene-specific boxplots are powered by Plotly. This dual approach allows users to hover over data points, zoom into regions of interest, and export results for further analysis.\n",
        "I implemented several quality-of-life functions, including intelligent file format detection for multiple input types (CSV, TSV, Excel, compressed files), error handling for malformed datasets, data caching for improved performance, and a modular code structure that separates data processing, statistical analysis, and visualization components. The result is a tool that makes differential expression analysis accessible while maintaining the statistical rigor required for high quality results."
      ],
      bullets: [],
    },
  },
  {
    id: "nanopore-basecalling",
    link: "/blog/nanopore-basecalling",
    companyName: "HMM Nanopore Basecaller",
    type: "Personal Project",
    category: ["Bioinformatics", "Computational Bioengineering","Blog"],
    shortDescription:
      "Explored and developed Hidden Markov Models for basecalling raw nanopore POD5 files.",
    websiteLink: "https://github.com/aahmeddev/nanopore-sequencing/",
    techStack: ["Python", "Biopython"],
    startDate: new Date("2025-07-29"),
    endDate: new Date("2025-08-15"),
    companyLogoImg: "/portfolio/projects/nanopore-basecalling/img.png",
    pagesInfoTitle: "",
    pagesInfoArr: [],
    descriptionDetails: {
      paragraphs: [],
      bullets: [],
    },
  }
];

export const featuredExperiences = Experiences.slice(0, 3);
